File: day1/day1.py
with open("input.txt") as f:
    curr = 50
    password = 0
    for line in f:
        if line[0] == "R":
            new_val = curr + int(line[1:])
            password += new_val // 100
            curr = new_val % 100
        else:
            new_val = curr - int(line[1:])
            if new_val <= 0:
                password += (1 if curr != 0 else 0) + (-new_val) // 100
            curr = new_val % 100
        print(curr, password, line)
    print(password)

File: day1/day1.hs
import Control.Monad
import System.IO

processLine :: (Int, Int) -> String -> IO (Int, Int)
processLine (curr, password) line = do
  let operation = head line
      value = read (tail line) :: Int
      (newCurr, newPassword) =
        if operation == 'R'
          then
            let newVal = curr + value
                addPassword = newVal `div` 100
                newCurrVal = newVal `mod` 100
             in (newCurrVal, password + addPassword)
          else
            let newVal = curr - value
                addPassword =
                  if newVal <= 0
                    then (if curr /= 0 then 1 else 0) + ((-newVal) `div` 100)
                    else 0
                newCurrVal = newVal `mod` 100
             in (newCurrVal, password + addPassword)

  putStrLn $ show newCurr ++ " " ++ show newPassword ++ " " ++ line
  return (newCurr, newPassword)

main :: IO ()
main = do
  contents <- readFile "input.txt"
  let linesOfFile = lines contents
  (_, finalPassword) <- foldM processLine (50, 0) linesOfFile
  print finalPassword
